packages/check/match.js

```diff
@@ -287,30 +320,44 @@ var checkSubtree = function (value, pattern) {
       requiredPatterns[key] = subPattern;
   });
 
-  _.each(value, function (subValue, key) {
-    try {
-      if (_.has(requiredPatterns, key)) {
-        checkSubtree(subValue, requiredPatterns[key]);
-        delete requiredPatterns[key];
-      } else if (_.has(optionalPatterns, key)) {
-        checkSubtree(subValue, optionalPatterns[key]);
-      } else {
-        if (!unknownKeysAllowed)
-          throw new Match.Error("Unknown key");
-        if (unknownKeyPattern) {
-          checkSubtree(subValue, unknownKeyPattern[0]);
+  for (var keys = _.keys(value), i = 0, length = keys.length; i < length; i++) {
+    var key = keys[i];
+    var subValue = value[key];
+    if (_.has(requiredPatterns, key)) {
+      var result = testSubtree(subValue, requiredPatterns[key]);
+      if (result) {
+        result.path = _prependPath(key, result.path);
+        return result;
+      }
+      delete requiredPatterns[key];
+    } else if (_.has(optionalPatterns, key)) {
+      var result = testSubtree(subValue, optionalPatterns[key]);
+      if (result) {
+        result.path = _prependPath(key, result.path);
+        return result;
+      }
+    } else {
+      if (!unknownKeysAllowed)
+        return {
+          message: "Unknown key",
+          path: key
+        };
+      if (unknownKeyPattern) {
+        var result = testSubtree(subValue, unknownKeyPattern[0]);
+        if (result) {
+          result.path = _prependPath(key, result.path);
+          return result;
         }
       }
-    } catch (err) {
-      if (err instanceof Match.Error)
-        err.path = _prependPath(key, err.path);
-      throw err;
     }
-  });
+  }
 
-  _.each(requiredPatterns, function (subPattern, key) {
-    throw new Match.Error("Missing key '" + key + "'");
-  });
+  for (var keys = _.keys(requiredPatterns), i = 0, length = keys.length; i < length; i++) {
```

Comment:
This doesn't actually need to be a loop, right? It will return on the first iteration, so we may as well do that in the code explicitly (I know the old code had a loop, but we should fix this now that we have the chance)